core/src/main/java/jenkins/telemetry/impl/java11/MissingClassTelemetry.java

```diff
@@ -0,0 +1,283 @@
+     * @return the map of missed classes events gathered along this window of java 11 telemetry
+     */
+    @Nonnull
+    private JSONArray formatEventsAndInitialize() {
+        // Save the current events and clean for next (not this one) telemetry send
+        ConcurrentHashMap<List<StackTraceElement>, MissingClassEvent> toReport = MissingClassTelemetry.events.getEventsAndClean();
+        if (LOGGER.isLoggable(Level.FINE)) LOGGER.fine("Cleaned events for missed classes on telemetry for Java 11");
+
+        return formatEvents(toReport);
+    }
+
+    /**
+     * Format the events gathered in a map used to create the json object to send via telemetry. The events are named by
+     * the class not found. But a class could be repeated if it was thrown from several places. The interesting
+     * pieces of information we want to gather are the places where the {@link ClassNotFoundException} or the
+     * {@link NoClassDefFoundError} errors happens, rather than the class itself.
+     * @param events events collected in this telemetry window.
+     * @return the events formatted in a map.
+     */
+    @Nonnull
+    private JSONArray formatEvents(@Nonnull ConcurrentHashMap<List<StackTraceElement>, MissingClassEvent> events) {
+        JSONArray jsonEvents = new JSONArray();
+
+        events.forEach((stackTrace, event) -> {
+            JSONObject eventObject = new JSONObject();
+            eventObject.put("className", event.getClassName());
+            eventObject.put("class", event.getClassName());
+            eventObject.put("time", event.getTime());
+            eventObject.put("occurrences", Long.toString(event.getOccurrences()));
+            eventObject.put("stacktrace", event.getStackTrace());
+
+            jsonEvents.add(eventObject);
+        });
+
+        return jsonEvents;
+    }
+
+    /**
+     * The current time in the same way as other telemetry implementations.
+     * @return the UTC time formatted with the pattern yyyy-MM-dd'T'HH:mm'Z'
+     */
+    @Nonnull
+    static String clientDateString() {
+        TimeZone tz = TimeZone.getTimeZone("UTC");
+        DateFormat df = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm'Z'");
+        df.setTimeZone(tz); // strip timezone
+        return df.format(new Date());
+    }
+
+    /**
+     * Store the exception if it's from a split package of Java. This method report this exception directly, it doesn't
+     * look into the causes or suppressed exceptions of the exception specified. This method tends to be used in the
+     * ClassLoader directly. Outside the class loaders is best to use {@link #reportExceptionInside(Throwable)}
+     * @param name the name of the class
+     * @param e the throwable to report if needed
+     */
+    public static void reportException(@Nonnull String name, @Nonnull Throwable e) {
+        //ClassDefFoundError uses / instead of .
+        name = name.replace('/', '.').trim();
+
+        if (isFromMovedPackage(name)) {
```

Comment:
It limits the usability of the telemetry to the code which was removed from Java.
So we won't be able to discover any issues related to libraries in Jenkins plugins

Reply:
I've been looking into the CNFE thrown during startup and after creating two jobs (freestyle and pipeline). **All of them are caught and ignored later in the code**. I've **added a check to avoid reporting** in such places. But as the check is **performance-expensive, it's done after** checking that the class is related with Java updates. I've reviewed a thousand cases, but **we can also have more cases that may give us false positives**. The check is: https://github.com/MRamonLeon/jenkins/blob/7f225169bdcc37e253eff81729536cd9efe294b7/core/src/main/java/jenkins/telemetry/impl/java11/MissingClassTelemetry.java#L233 using this list of ignored places: https://github.com/MRamonLeon/jenkins/blob/7f225169bdcc37e253eff81729536cd9efe294b7/core/src/main/java/jenkins/telemetry/impl/java11/MissingClassTelemetry.java#L92